File: code/chap01/rdd_transformation_cartesian.py
#!/usr/bin/python
#-----------------------------------------------------
# Apply a cartesian() transformation to an RDD
# Input: NONE
#------------------------------------------------------
# Input Parameters:
#    NONE
#-------------------------------------------------------
# @author Mahmoud Parsian
#-------------------------------------------------------
from __future__ import print_function 
import sys 
from pyspark.sql import SparkSession 

#========================================
def main():

    # create an instance of SparkSession
    spark = SparkSession.builder.getOrCreate()


    #========================================
    # cartesian() transformation
    #
    # cartesian(other)
    # Return the Cartesian product of this RDD 
    # and another one, that is, the RDD of all 
    # pairs of elements (a, b) where a is in 
    # self (as source-one RDD) and b is in other
    # (as source-two RDD).
    #
    #========================================
    a = [('a', 2), ('b', 3), ('c', 4)]
    print("a = ", a)
    #
    b = [('p', 50), ('x', 60), ('y', 70), ('z', 80) ]
    print("b = ", b)
    #
    rdd = spark.sparkContext.parallelize(a)
    print("rdd = ", rdd)
    print("rdd.count() = ", rdd.count())
    print("rdd.collect() = ", rdd.collect())
    #
    rdd2 = spark.sparkContext.parallelize(b)
    print("rdd2 = ", rdd2)
    print("rdd2.count() = ", rdd2.count())
    print("rdd2.collect() = ", rdd2.collect())
    
    #
    cart = rdd.cartesian(rdd2)
    print("cart = ", cart)
    print("cart.count() = ", cart.count())
    print("cart.collect() = ", cart.collect())
    
    
    # done!
    spark.stop()
#end-def
#==================================
if __name__ == '__main__':
    main()

File: code/chap01/scala/src/main/scala/org/data/algorithms/spark/ch01/RDDTransformationCartesian.scala
package org.data.algorithms.spark.ch01

import org.apache.spark.sql.SparkSession

/**
 *-------------------------------------------------------
 * Apply a cartesian() transformation to an RDD
 *-------------------------------------------------------
 *
 * @author Biman Mandal
 *-------------------------------------------------------
 */
object RDDTransformationCartesian {

  def main(args: Array[String]): Unit = {

    // create an instance of SparkSession
    val spark = SparkSession.builder().master("local[*]").getOrCreate()

    /**
     * ----------------------------------------------------
     * cartesian() transformation
     *
     * cartesian(other)
     * Return the Cartesian product of this RDD
     * and another one, that is, the RDD of all
     * pairs of elements (a, b) where a is in
     * self (as source-one RDD) and b is in other
     * (as source-two RDD).
     *
     * ----------------------------------------------------
     */
    val a = List(("a", 2), ("b", 3), ("c", 4))
    println("a = " + a)

    val b = List(("p", 50), ("x", 60), ("y", 70), ("z", 80) )
    println("b = " + b)

    val rdd = spark.sparkContext.parallelize(a)
    println("rdd = " + rdd)
    println("rdd.count() = " + rdd.count())
    println("rdd.collect() = " + rdd.collect().mkString("Array(", ", ", ")"))

    val rdd2 = spark.sparkContext.parallelize(b)
    println("rdd2 = " + rdd2)
    println("rdd2.count() = " + rdd2.count())
    println("rdd2.collect() = " + rdd2.collect().mkString("Array(", ", ", ")"))

    val cart = rdd.cartesian(rdd2)
    println("cart = " + cart)
    println("cart.count() = " + cart.count())
    println("cart.collect() = " + cart.collect().mkString("Array(", ", ", ")"))

    // done!
    spark.stop()

  }

}
